File: generated-code/go/codecraft/TcpReadWrite/model/debug_interface/colored_vertex.go
package debug_interface

import "io"
import . "trans_gen_test/common"
import . "trans_gen_test/stream"

// Vertex for debug rendering
type ColoredVertex struct {
    // Position in world coordinates (if none, screen position (0, 0) is used)
    WorldPos *Vec2Float32
    // Additional offset in screen coordinates
    ScreenOffset Vec2Float32
    // Color to use
    Color Color
}

func NewColoredVertex(worldPos *Vec2Float32, screenOffset Vec2Float32, color Color) ColoredVertex {
    return ColoredVertex {
        WorldPos: worldPos,
        ScreenOffset: screenOffset,
        Color: color,
    }
}

// Read ColoredVertex from reader
func ReadColoredVertex(reader io.Reader) ColoredVertex {
    var worldPos *Vec2Float32
    if ReadBool(reader) {
        var worldPosValue Vec2Float32
        worldPosValue = ReadVec2Float32(reader)
        worldPos = &worldPosValue
    } else {
        worldPos = nil
    }
    var screenOffset Vec2Float32
    screenOffset = ReadVec2Float32(reader)
    var color Color
    color = ReadColor(reader)
    return ColoredVertex {
        WorldPos: worldPos,
        ScreenOffset: screenOffset,
        Color: color,
    }
}

// Write ColoredVertex to writer
func (coloredVertex ColoredVertex) Write(writer io.Writer) {
    worldPos := coloredVertex.WorldPos
    if worldPos == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        worldPosValue := *worldPos
        worldPosValue.Write(writer)
    }
    screenOffset := coloredVertex.ScreenOffset
    screenOffset.Write(writer)
    color := coloredVertex.Color
    color.Write(writer)
}

// Get string representation of ColoredVertex
func (coloredVertex ColoredVertex) String() string {
    stringResult := "{ "
    stringResult += "WorldPos: "
    worldPos := coloredVertex.WorldPos
    if worldPos == nil {
        stringResult += "nil"
    } else {
        worldPosValue := *worldPos
        stringResult += worldPosValue.String()
    }
    stringResult += ", "
    stringResult += "ScreenOffset: "
    screenOffset := coloredVertex.ScreenOffset
    stringResult += screenOffset.String()
    stringResult += ", "
    stringResult += "Color: "
    color := coloredVertex.Color
    stringResult += color.String()
    stringResult += " }"
    return stringResult
}

File: generated-code/javascript/codecraft/FileReadWrite/model/debug-interface/colored-vertex.js
const Color = require.main.require('./color');
const Vec2Float = require.main.require('./vec2-float');
/**
 * Vertex for debug rendering
 */
class ColoredVertex {
    /**
     * Position in world coordinates (if none, screen position (0, 0) is used)
     */
    worldPos;
    /**
     * Additional offset in screen coordinates
     */
    screenOffset;
    /**
     * Color to use
     */
    color;

    constructor(worldPos, screenOffset, color) {
        this.worldPos = worldPos;
        this.screenOffset = screenOffset;
        this.color = color;
    }

    /**
     * Read ColoredVertex from input stream
     */
    static async readFrom(stream) {
        let worldPos;
        if (await stream.readBool()) {
            worldPos = await Vec2Float.readFrom(stream);
        } else {
            worldPos = null;
        }
        let screenOffset;
        screenOffset = await Vec2Float.readFrom(stream);
        let color;
        color = await Color.readFrom(stream);
        return new ColoredVertex(worldPos, screenOffset, color);
    }

    /**
     * Write ColoredVertex to output stream
     */
    async writeTo(stream) {
        let worldPos = this.worldPos;
        if (worldPos === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await worldPos.writeTo(stream);
        }
        let screenOffset = this.screenOffset;
        await screenOffset.writeTo(stream);
        let color = this.color;
        await color.writeTo(stream);
    }
}
module.exports = ColoredVertex